rd_ui/app/scripts/embed.js

```diff
@@ -43,12 +43,21 @@ angular.module('redash', [
     }
   ])
    .controller('EmbedCtrl', ['$scope', function ($scope) {} ])
-   .controller('EmbeddedVisualizationCtrl', ['$scope', 'Query', 'QueryResult',
-    function ($scope, Query, QueryResult) {
+   .controller('EmbeddedVisualizationCtrl', ['$scope', '$location', 'Query', 'QueryResult',
+    function ($scope, $location, Query, QueryResult) {
        $scope.embed = true;
        $scope.visualization = visualization;
        $scope.query = visualization.query;
        query = new Query(visualization.query);
-       $scope.queryResult = new QueryResult({query_result:query_result});
+
+       //max age from querystring, default to -1
+       maxAge = $location.search()['maxAge'];
+       if (maxAge === undefined) {
+          maxAge = -1;
+       }
+
+       qr = query.getQueryResult(maxAge, Query.collectParamsFromQueryString($location, query));
```

Comment:
Note that you don't use the `query_result` object, which means:
1. This won't work for unautheticated users (the call to query result will be without the API key).
2. No point in passing it or preloading it.

I think your direction is the right one, but needs a bit more work to work in all scenarios.

My suggestion is:
Rewrite the [`embed`](https://github.com/getredash/redash/blob/master/redash/handlers/embed.py#L19) endpoint to be more like [`public_dashboard`](https://github.com/getredash/redash/blob/master/redash/handlers/embed.py#L52) and use the same template (`public.html`). We can drop the `embed.js` file and move the routes definitions into `app.js`. 

Pass in the `seed_data` only the query object, but use the API to receive the query result (either a new one or existing one with the query result id). In app.js we already [take care of passing the API key](https://github.com/getredash/redash/blob/master/rd_ui/app/scripts/app.js#L28) to subsequent API calls so it will work both for signed in and signed out users.